### ISSUE TITLE
Question regarding BucketIterator

### ISSUE THREAD

**Issue by uccollab** [bentrevett/pytorch-sentiment-analysis#45]
hi, I'm apporaching sentiment analysis with torchtext and I've recently been studying the concept of Iterator. From what I understand it is used to automatically convert strings in vectors, batching them (that is, getting the set of vectors that shall be used for training) and then move them to the computing device.

I saw that BucketIterator tries to get a batch in which all the sentences have similar length, to reduce the amount of padding. My question is: if a sentence is shorter than the fixed length it is padded, but what if a sentence is longer? Is it truncated? If yes, how exactly?

Thanks in advance.

**Comment by bentrevett** [comment]
An `Iterator` will sample a batch of sentences, each will (usually) have different lengths, say: 32, 25, 10 tokens in each sentence. It will then look at the longest sentence within the batch, here 32 tokens, and then all sentences shorter than that will be padded to that length, 32 tokens. This will be done by appending a `pad_token`, which is `<pad>` by default, to the end of the sentences.

The problem with the standard  `Iterator` is that it creates batches by sampling sentences from the dataset randomly. This means we'll probably have batches of sentences with wildly different lengths and thus some sentences will have a lot of padding, e.g. in the above example the sentence with 10 tokens will have 22 padding tokens appended to the end of it. If we're using an RNN/LSTM type model, this means the last 22 time-steps on that sentence will just be over padding tokens - which is wasted computation.

The `BucketIterator` helps with this problem by trying to give you batches where each sentence is roughly the same length, i.e. a batch of lengths 32, 30, 28 is more likely than a batch of 32, 16, 10. This cuts down on the amount of padding within the batch, which means less computation over padding tokens and leads to less computation time overall. 

The exact way the `BucketIterator` works is given in the code [here](https://github.com/pytorch/text/blob/master/torchtext/data/iterator.py#L274). It buckets your dataset into buckets of size `100 * batch size`, sorts each bucket by their length (using the `sort_key`), then cuts each bucket into 100 batches of `batch_size` each, then yields a random batch from a random bucket. These random batches have been sorted by their length, so usually have similar sentence lengths and thus less padding.

I'm not sure what you mean by "fixed length" in your question above. There is no set fixed length to pad to, padding is always done so all sentences within the batch are equal to the longest sentence.

**Comment by uccollab** [comment]
Ok now I understand this. Than you for the time spent explaining it!

**Comment by maroxtn** [comment]
Is there a way for having the same feature on a regular DataLoader ?
